File: services/claims_scoring/src/kafka_publisher.rs
use crate::models::{ClaimDecisionEvent, ClaimStatus};
use anyhow::Result;
use rdkafka::config::ClientConfig;
use rdkafka::producer::{FutureProducer, FutureRecord};
use std::time::Duration;
use tracing::{error, info};

/// Publishes a `claim_decision` event to Kafka.
///
/// On AUTO_APPROVED, also publishes a `payout_authorized` event.
/// On FRAUD_QUEUE, also publishes a `fraud_alert` event.
pub async fn publish_claim_decision(event: &ClaimDecisionEvent, status: &ClaimStatus) -> Result<()> {
    let brokers = std::env::var("KAFKA_BROKERS").unwrap_or_else(|_| "localhost:9092".to_string());

    let producer: FutureProducer = ClientConfig::new()
        .set("bootstrap.servers", &brokers)
        .set("message.timeout.ms", "5000")
        .create()?;

    // Publish claim_decision event
    let payload = serde_json::to_string(event)?;
    let key = event.claim_id.to_string();

    let delivery_result = producer
        .send(
            FutureRecord::to("claim_decision")
                .payload(&payload)
                .key(&key),
            Duration::from_secs(5),
        )
        .await;

    match delivery_result {
        Ok((partition, offset)) => {
            info!(
                claim_id = %event.claim_id,
                partition,
                offset,
                "Published claim_decision to Kafka"
            );
        }
        Err((e, _)) => {
            error!(
                claim_id = %event.claim_id,
                error = %e,
                "Failed to publish claim_decision to Kafka"
            );
            return Err(anyhow::anyhow!("Kafka publish failed: {e}"));
        }
    }

    // Publish secondary event based on routing decision
    match status {
        ClaimStatus::AutoApproved => {
            publish_payout_authorized(&producer, event).await?;
        }
        ClaimStatus::FraudQueue => {
            publish_fraud_alert(&producer, event).await?;
        }
        _ => {
            // DEVICE_NOT_ATTESTED and PLATFORM_ACTIVITY_VETO do not trigger secondary events
        }
    }

    Ok(())
}

async fn publish_payout_authorized(producer: &FutureProducer, event: &ClaimDecisionEvent) -> Result<()> {
    let payload = serde_json::json!({
        "claim_id": event.claim_id,
        "worker_id": event.worker_id,
        "fraud_score": event.fraud_score,
        "authorized_at": event.decided_at,
    });

    let payload_str = serde_json::to_string(&payload)?;
    let key = event.claim_id.to_string();

    let delivery_result = producer
        .send(
            FutureRecord::to("payout_authorized")
                .payload(&payload_str)
                .key(&key),
            Duration::from_secs(5),
        )
        .await;

    match delivery_result {
        Ok((partition, offset)) => {
            info!(
                claim_id = %event.claim_id,
                partition,
                offset,
                "Published payout_authorized to Kafka"
            );
            Ok(())
        }
        Err((e, _)) => {
            error!(
                claim_id = %event.claim_id,
                error = %e,
                "Failed to publish payout_authorized to Kafka"
            );
            Err(anyhow::anyhow!("Kafka publish failed: {e}"))
        }
    }
}

async fn publish_fraud_alert(producer: &FutureProducer, event: &ClaimDecisionEvent) -> Result<()> {
    let payload = serde_json::json!({
        "alert_type": "fraud_queue_routing",
        "claim_id": event.claim_id,
        "worker_id": event.worker_id,
        "fraud_score": event.fraud_score,
        "triggered_at": event.decided_at,
    });

    let payload_str = serde_json::to_string(&payload)?;
    let key = event.claim_id.to_string();

    let delivery_result = producer
        .send(
            FutureRecord::to("fraud_alert")
                .payload(&payload_str)
                .key(&key),
            Duration::from_secs(5),
        )
        .await;

    match delivery_result {
        Ok((partition, offset)) => {
            info!(
                claim_id = %event.claim_id,
                partition,
                offset,
                "Published fraud_alert to Kafka"
            );
            Ok(())
        }
        Err((e, _)) => {
            error!(
                claim_id = %event.claim_id,
                error = %e,
                "Failed to publish fraud_alert to Kafka"
            );
            Err(anyhow::anyhow!("Kafka publish failed: {e}"))
        }
    }
}

File: services/claims_scoring/kafka_publisher.py
import json
import logging
import os

logger = logging.getLogger(__name__)


async def publish_claim_decision(response, worker_id: str) -> None:
    try:
        from aiokafka import AIOKafkaProducer
    except ImportError:
        logger.warning("aiokafka not available — skipping Kafka publish")
        return

    from models import ClaimStatus

    brokers = os.getenv("KAFKA_BROKERS", "localhost:9092")
    event = {
        "claim_id": str(response.claim_id),
        "worker_id": str(worker_id),
        "status": response.status.value.lower(),
        "fraud_score": response.fraud_score,
        "decided_at": response.decided_at.isoformat(),
    }

    producer = AIOKafkaProducer(bootstrap_servers=brokers)
    try:
        await producer.start()
        payload = json.dumps(event).encode()
        key = str(response.claim_id).encode()

        await producer.send_and_wait("claim_decision", value=payload, key=key)
        logger.info("Published claim_decision claim_id=%s", response.claim_id)

        if response.status == ClaimStatus.AUTO_APPROVED:
            payout_event = {
                "claim_id": str(response.claim_id),
                "worker_id": str(worker_id),
                "fraud_score": response.fraud_score,
                "authorized_at": response.decided_at.isoformat(),
            }
            await producer.send_and_wait(
                "payout_authorized",
                value=json.dumps(payout_event).encode(),
                key=key,
            )
            logger.info("Published payout_authorized claim_id=%s", response.claim_id)

        elif response.status == ClaimStatus.FRAUD_QUEUE:
            alert_event = {
                "alert_type": "fraud_queue_routing",
                "claim_id": str(response.claim_id),
                "worker_id": str(worker_id),
                "fraud_score": response.fraud_score,
                "triggered_at": response.decided_at.isoformat(),
            }
            await producer.send_and_wait(
                "fraud_alert",
                value=json.dumps(alert_event).encode(),
                key=key,
            )
            logger.info("Published fraud_alert claim_id=%s", response.claim_id)
    finally:
        await producer.stop()
